vowpalwabbit/array_parameters_dense.h

```diff
@@ -85,7 +85,11 @@ class dense_parameters
   const_iterator cend() { return const_iterator(_begin + _weight_mask + 1, _begin, stride()); }
 
   inline const weight& operator[](size_t i) const { return _begin[i & _weight_mask]; }
-  inline weight& operator[](size_t i) { return _begin[i & _weight_mask]; }
+  inline weight& operator[](size_t i)
+  {
+    if (_begin == nullptr) { THROW("Error: Dense parameters not initialized."); }
```

Comment:
Putting a throw in this operator could slow things down quite a bit. Consider making it an assert? Can this condition be handled somewhere else in a sort of validation step?

Reply:
Yes there are a few places I can add this check. This is being invoked by this function:
`inline weight& strided_index(size_t index) { return operator[](index << _stride_shift); }`
I think it make sense to assert `not_null()` within the strided_index function. Alternatively I could check `not_null()` at an earlier stage, but that might leave future errors.

Reply:
Generally for a function like `operator[]` you define under what conditions it is undefined behavior. We can assert that state in a debug build. I feel like the initialized check should be done once after the model is loaded?

Reply:
Ok yea I changed the check to when the model is loaded